materialize-snowflake/snowflake.go

```diff
@@ -596,93 +609,177 @@ func (d *transactor) Load(it *pm.LoadIterator, loaded func(int, json.RawMessage)
-		err = fmt.Errorf("txn.Exec: %w", err)
+// applyCheckpoint merges data from temporary table to main table
+func (d *transactor) applyCheckpoint(ctx context.Context, cp checkpoint) error {
+	var asyncCtx = sf.WithAsyncMode(ctx)
+	log.Info("store: starting committing changes")
+
+	// Run the queries using AsyncMode, which means that `ExecContext` will not block
+	// until the query is successful, rather we will store the results of these queries
+	// in a map so that we can then call `RowsAffected` on them, blocking until
+	// the queries are actually executed and done
+	var results = make(map[string]stdsql.Result)
+	for table, queries := range cp.Queries {
+		// during recovery we skip queries that belong to tables which do not have a binding anymore
+		// since these tables might be deleted already
+		if !d.hasTableBinding(table) {
+			continue
+		}
 
-		// Give recommendation to user for resolving timeout issues
-		if strings.Contains(err.Error(), "timeout") {
-			return fmt.Errorf("fence.Update: %w  (ensure LOCK_TIMEOUT and STATEMENT_TIMEOUT_IN_SECONDS are at least ten minutes)", err)
+		for _, q := range queries {
+			if result, err := d.store.conn.ExecContext(asyncCtx, q); err != nil {
+				return fmt.Errorf("query %q failed: %w", q, err)
+			} else {
+				results[q] = result
+			}
 		}
+	}
 
-		return err
+	for q, r := range results {
+		if _, err := r.RowsAffected(); err != nil {
+			return fmt.Errorf("query %q failed: %w", q, err)
+		}
 	}
 
-	log.Info("store: starting copying of files into tables")
-	for _, b := range d.bindings {
-		if !b.store.stage.started {
-			// No table update required
-		} else if err := b.store.stage.flush(); err != nil {
-			return err
-		} else if !b.store.mustMerge {
-			log.WithField("table", b.target.Identifier).Info("store: starting direct copying data into table")
-			// We can issue a faster COPY INTO the target table.
-			if _, err = txn.ExecContext(ctx, b.store.copyInto); err != nil {
-				return fmt.Errorf("copying Store documents into table %q: %w", b.target.Identifier, err)
-			}
-			log.WithField("table", b.target.Identifier).Info("store: finishing direct copying data into table")
-		} else {
-			log.WithField("table", b.target.Identifier).Info("store: starting merging data into table")
-			// We must MERGE into the target table.
-			if _, err = txn.ExecContext(ctx, b.store.mergeInto); err != nil {
-				return fmt.Errorf("merging Store documents into table %q: %w", b.target.Identifier, err)
-			}
-			log.WithField("table", b.target.Identifier).Info("store: finishing merging data into table")
+	for table, toDelete := range cp.ToDelete {
+		// during recovery we skip queries that belong to tables which do not have a binding anymore
+		// since these tables might be deleted already
+		if !d.hasTableBinding(table) {
+			continue
 		}
 
-		// Reset for next transaction.
-		b.store.mustMerge = false
+		d.deleteFiles(ctx, toDelete)
```

Comment:
What happens if applyCheckpoint runs to completion, but we fail to persist the connector state update, and it's run again? Does it matter that files are deleted? I assume not but just confirming.

Reply:
It's not a problem, Snowflake will ignore nonexistent files so long as they are not explicitly listed in a `FILES` clause. In our case we ask Snowflake to copy from a directory.